query.go

```diff
@@ -293,78 +312,116 @@ func (q *query) run() {
-		// Note: NumWaiting will be updated in spawnQuery
-		maxNumQueriesToSpawn := alpha - q.queryPeers.NumWaiting()
 		// try spawning the queries, if there are no available peers to query then we won't spawn them
-		for j := 0; j < maxNumQueriesToSpawn; j++ {
-			q.spawnQuery(pathCtx, cause, ch)
+		for _, p := range qPeers {
+			q.spawnQuery(pathCtx, cause, p, ch)
 		}
 	}
 }
 
 // spawnQuery starts one query, if an available heard peer is found
-func (q *query) spawnQuery(ctx context.Context, cause peer.ID, ch chan<- *queryUpdate) {
-	peers := q.queryPeers.GetSortedHeard()
-	if len(peers) == 0 {
-		return
-	}
-
+func (q *query) spawnQuery(ctx context.Context, cause peer.ID, queryPeer peer.ID, ch chan<- *queryUpdate) {
 	PublishLookupEvent(ctx,
 		NewLookupEvent(
 			q.dht.self,
 			q.id,
 			q.key,
 			NewLookupUpdateEvent(
 				cause,
-				q.queryPeers.GetReferrer(peers[0]),
-				nil,                 // heard
-				[]peer.ID{peers[0]}, // waiting
-				nil,                 // queried
-				nil,                 // unreachable
+				q.queryPeers.GetReferrer(queryPeer),
+				nil,                  // heard
+				[]peer.ID{queryPeer}, // waiting
+				nil,                  // queried
+				nil,                  // unreachable
+				nil,                  // rejected
 			),
 			nil,
 			nil,
 		),
 	)
-	q.queryPeers.SetState(peers[0], qpeerset.PeerWaiting)
+	q.queryPeers.SetState(queryPeer, qpeerset.PeerWaiting)
 	q.waitGroup.Add(1)
-	go q.queryPeer(ctx, ch, peers[0])
+	go q.queryPeer(ctx, ch, queryPeer)
 }
 
-func (q *query) isReadyToTerminate() (bool, LookupTerminationReason) {
+func (q *query) isReadyToTerminate(ctx context.Context, nPeersToQuery int) (bool, LookupTerminationReason, []peer.ID) {
 	// give the application logic a chance to terminate
 	if q.stopFn() {
-		return true, LookupStopped
+		return true, LookupStopped, nil
 	}
 	if q.isStarvationTermination() {
-		return true, LookupStarvation
+		return true, LookupStarvation, nil
 	}
 	if q.isLookupTermination() {
-		return true, LookupCompleted
+		return true, LookupCompleted, nil
 	}
-	return false, -1
+
+	// The peers we query next should be ones that are not rejected by the filter.
+	var peersToQuery []peer.ID
+	peers := q.queryPeers.GetClosestInStates(qpeerset.PeerHeard)
+	for _, p := range peers {
+		if q.dFilter != nil {
+			allowed := q.dFilter.TryAdd(p)
+			if !allowed {
+				src := q.queryPeers.GetReferrer(p)
+				PublishLookupEvent(ctx,
+					NewLookupEvent(
+						q.dht.self,
+						q.id,
+						q.key,
+						NewLookupUpdateEvent(
+							src,
+							src,
+							nil,          // heard
+							nil,          // waiting
+							nil,          // queried
+							nil,          // unreachable
+							[]peer.ID{p}, // rejected
+						),
+						nil,
+						nil,
+					),
+				)
+				q.queryPeers.SetState(p, qpeerset.PeerRejected)
+				continue
+			}
+		}
+		peersToQuery = append(peersToQuery, p)
+	}
+
+	if len(peersToQuery) >= nPeersToQuery {
+		peersToQuery = peersToQuery[:nPeersToQuery]
```

Comment:
When you cut down the slice, for all cut peers, you also need to undo the adding to the filter and the state setting. I think the easiest thing to do is to limit the for-loop itself to not exceed nPeersToQuery steps.

Reply:
Thanks for this catch ! Have fixed this.